### ISSUE TITLE
tensor3_from_bytes() scaling parameters 

### ISSUE THREAD

**Issue by igorgya** [Dobiasd/frugally-deep#78]
I preprocess an OpenCV  Mat image before I call tensor3_from_bytes() function to convert it to the tensor3.
Is there the way to not use the scaling parameters in that function?

**Comment by Dobiasd** [comment]
Hi @igorgya 

sorry for the late response. Since you closed the issue, I guess you already found the solution.

For future reference however:

[`fdeep::tensor3_from_bytes`](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/tensor3.hpp#L568) takes byte values in the range of [0, 255]. By default it converts to floats in the range [0.0, 1.0].
The target range can be changed by adjusting its two parameters `low` and `high`. E.g., the following will leave the value range as it is in the input:

```cpp
fdeep::tensor3_from_bytes(some_byte_ptr, height, width, channel_count, 0.0, 255.0);
```

**Comment by igorgya** [comment]
Tobias, thank you for your response.
I have retrained my cnn with 0.0 -1.0 range and it works now the same way as with python. Your frugally-deep is one great library!

Get Outlook for iOS<https://aka.ms/o0ukef>



On Tue, Aug 14, 2018 at 11:29 AM -0400, "Tobias Hermann" <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @igorgya<https://github.com/igorgya>

sorry for the late response. Since you closed the issue, I guess you already found the solution.

For future reference however:

fdeep::tensor3_from_bytes<https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/tensor3.hpp#L568> takes byte values in the range of [0, 255]. By default it converts to floats in the range [0.0, 1.0].
The target range can be changed by adjusting its two parameters low and high. E.g., the following will leave the value range as it is in the input:

fdeep::tensor3_from_bytes(some_byte_ptr, height, width, channel_count, 0.0, 255.0);

**Comment by Dobiasd** [comment]
Thank you. :slightly_smiling_face: 

Do you think the [documentation](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/tensor3.hpp#L564) of `fdeep::tensor3_from_bytes` or its [parameter names](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/tensor3.hpp#L570) could be improved, so that for future users this thing is more obvious?

**Comment by igorgya** [comment]
I don't think so. The documentation explicitly states input parameters to fdeep::tensor3_from_bytes().
Somehow I missed that and rushed for the easy answer!

**Comment by igorgya** [comment]
Tobias,
Sorry to bother you over email instead of go thru GitHub. I have a different problem now.
My model is a modified UNET(Upsampling2D + Conv2D instead of all Conv2DTranspose) and although
The results of frugally-deep are the same as with Keras there is a significant difference between Keras run on CPU and VS 2015 release frugally-deep build: 0.65s vs 4.6s.
I wanted to upload my json and h5 files along with minimum code in zip format on GitHub but it was 50MB and not accepted.

Regards,
Igor

**Comment by Dobiasd** [comment]
You wrote your question to the GitHub issue, not as a private email. :wink: 

Anyway, please make sure you are compiling your code in Visual Studio with performance optimizations switched on, i.e., are you building and running a Release executable and not a Debug one.

**Comment by igorgya** [comment]
In Debug mode it takes > 400s
